### PR TITLE
added values for config parser

### PR DESCRIPTION
**Pull request by SahithiMy**
### :nut_and_bolt: Description: What code changed, and why?
Added the structure for GCS fields and validated those fields for config parser.

### :chains: Related Resources
https://chefio.atlassian.net/browse/CHEF-5609
https://chefio.atlassian.net/browse/CHEF-5610
### :+1: Definition of Done

- Added the structure for Config parser.

- Validated the fields of GCS in config parser.

### :athletic_shoe: How to Build and Test the Change
`rebuild components/automate-cli`

### :white_check_mark: Checklist

**All PRs** must tick these:

- [X] I have read the [CONTRIBUTING document](https://github.com/chef/automate/blob/master/CONTRIBUTING.md).
- [X] All commits signed-off for the [Developer Certification of Origin](https://github.com/chef/chef/blob/master/CONTRIBUTING.md#developer-certification-of-origin-dco).

With occasional exceptions, all PRs **from Progress employees** must tick these:

- [X] Is the code clear? *(complicated code or lots of comments--subdivide and use well-named methods, meaningful variable names, etc.)*
- [X] Consistency checked? *(user notifications, user prompts, visual patterns, code patterns, variable names)*
- [X] Repeated code blocks eliminated? *(adapt and reuse existing components, blocks, functions, etc.)*
- [X] Spelling, grammar, typos checked? *(at a minimum use `make spell` in any component directory)*
- [X] Code well-formatted? *(indents, line breaks, etc. improve rather than hinder readability)*

All PRs **from Progress employees** should tick these if appropriate:

- [X] Tests added/updated? (all new code needs new tests)
- [X] Docs added/updated? (all customer-facing changes)

*Please add a note next to any checkbox above if you are NOT ticking it.*

### :camera: Screenshots, if applicable
Screenshot 2023-08-24 at 2 03 47 PM

### CODE REVIEW

File: lib/config/hadeployconfig_test.go [hunk 1]
```diff
@@ -819,7 +819,107 @@ Dv6bUUXSsZF4fb1diLIBpmD1hh8OGNY65LUPpzAxJeZvo5w=
...
+						WorkspacePath:               "/hab/a2_deploy_workspace",
+						BackupMount:                 "automate_backups",
+						BackupConfig:                "object_storage",
+						S3BucketName:                "",
+						HabitatUIDGid:               "",
+					},
+				},
+				ObjectStorage: &ObjectStorage{
+					Config: &ConfigObjectStorage{
+						Location:       "gcs",
```
Comment (daveaugustus) [inline lib/config/hadeployconfig_test.go:847]:
```suggestion
						Location:       "GCS",
```

File: lib/config/hadeployconfig.go [hunk 2]
```diff
@@ -237,6 +255,21 @@ func (c *HaDeployConfig) Parse(configFile string) error {
 	if err != nil {
 		return fmt.Errorf("error unmarshalling config TOML file: %w", err)
 	}
+	gcpServiceAccount := GcpServiceAccount{}
+	if c.GetConfigInitials() != nil && c.GetConfigInitials().BackupConfig == "object_storage" {
+		objectStorageConfig := c.GetObjectStorageConfig()
+		if objectStorageConfig.Location == "gcs" {
```
Comment (daveaugustus) [inline lib/config/hadeployconfig.go:261]:
```suggestion
		if objectStorageConfig.Location == "GCS" {
```

File: lib/config/hadeployconfigverify.go [hunk 3]
```diff
@@ -174,22 +174,29 @@ func (c *HaDeployConfig) validateAwsBackupConfig() error {
...
-	if err := validateS3Endpoint(objectStorage.Endpoint); err != nil {
-		errorList.PushBack(err)
+	if objectStorage.Location == "gcs" {
+		if err := c.verifyGcsStorage(objectStorage.GcpServiceAccount); err != nil {
+			errorList.PushBack(err)
+		}
 	}
-	if objectStorage.Region != "" {
-		if err := validateS3AWSRegion(objectStorage.Region); err != nil {
+	if objectStorage.Location == "s3" {
```
Comment (daveaugustus) [inline lib/config/hadeployconfigverify.go:185]:
```suggestion
	if objectStorage.Location == "S3" {
```

File: lib/config/testdata/HaOnPremGcs.toml [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+# DON'T MODIFY THE BELOW LINE (backup_mount)
+backup_mount = "automate_backups"
+
+# Eg.: backup_config = "object_storage" or "file_system"
+backup_config = "object_storage"
+
+# If backup_config = "object_storage" fill out [object_storage.config] as well 
+## Object storage similar to AWS S3 Bucket
+[object_storage.config]
+location = "gcs"
```
Comment (daveaugustus) [inline lib/config/testdata/HaOnPremGcs.toml:35]:
```suggestion
location = "GCS"
```

File: lib/config/testdata/HaOnPremGcsFailed.toml [hunk 5]
```diff
@@ -0,0 +1,11 @@
+[architecture.existing_infra]
+# Eg.: backup_config = "object_storage" or "file_system"
+backup_config = "object_storage"
+
+# If backup_config = "object_storage" fill out [object_storage.config] as well 
+## Object storage similar to AWS S3 Bucket
+[object_storage.config]
+location = "gcs"
```
Comment (daveaugustus) [inline lib/config/testdata/HaOnPremGcsFailed.toml:8]:
```suggestion
location = "GCS"
```

File: lib/config/testdata/UnmarshalErrorForGcs.toml [hunk 6]
```diff
@@ -0,0 +1,15 @@
...
+
+[architecture.existing_infra]
+
+# Eg.: backup_config = "object_storage" or "file_system"
+backup_config = "object_storage"
+
+# If backup_config = "object_storage" fill out [object_storage.config] as well 
+## Object storage similar to AWS S3 Bucket
+[object_storage.config]
+location = "gcs"
```
Comment (daveaugustus) [inline lib/config/testdata/UnmarshalErrorForGcs.toml:13]:
```suggestion
location = "GCS"
```

File: lib/config/hadeployconfig_test.go [hunk 7]
```diff
@@ -833,6 +933,46 @@ Dv6bUUXSsZF4fb1diLIBpmD1hh8OGNY65LUPpzAxJeZvo5w=
...
+			args: args{configFile: "./testdata/HaOnPremGcsFailed.toml"},
+			want: &HaDeployConfig{
+				Architecture: &Architecture{
+					ExistingInfra: &ConfigInitials{
+						BackupConfig: "object_storage",
+					},
+				},
+				ObjectStorage: &ObjectStorage{
+					Config: &ConfigObjectStorage{
+						Location:       "gcs",
```
Comment (daveaugustus) [inline lib/config/hadeployconfig_test.go:947]:
```suggestion
						Location:       "GCS",
```

File: lib/config/hadeployconfig_test.go [hunk 8]
```diff
@@ -833,6 +933,46 @@ Dv6bUUXSsZF4fb1diLIBpmD1hh8OGNY65LUPpzAxJeZvo5w=
...
+			args: args{configFile: "./testdata/UnmarshalErrorForGcs.toml"},
+			want: &HaDeployConfig{
+				Architecture: &Architecture{
+					ExistingInfra: &ConfigInitials{
+						BackupConfig: "object_storage",
+					},
+				},
+				ObjectStorage: &ObjectStorage{
+					Config: &ConfigObjectStorage{
+						Location:       "gcs",
```
Comment (daveaugustus) [inline lib/config/hadeployconfig_test.go:967]:
```suggestion
						Location:       "GCS",
```

File: lib/config/hadeployconfigverify.go [hunk 9]
```diff
@@ -174,22 +174,29 @@ func (c *HaDeployConfig) validateAwsBackupConfig() error {
...
-	}
-	if err := validateRequiredString(objectStorage.SecretKey, "secret_key"); err != nil {
-		errorList.PushBack(err)
-	}
 	if err := validateRequiredString(objectStorage.BucketName, "bucket_name"); err != nil {
 		errorList.PushBack(err)
 	}
-	if err := validateS3Endpoint(objectStorage.Endpoint); err != nil {
-		errorList.PushBack(err)
+	if objectStorage.Location == "gcs" {
```
Comment (daveaugustus) [inline lib/config/hadeployconfigverify.go:180]:
```suggestion
	if objectStorage.Location == "GCS" {
```

File: lib/config/hadeployconfigverify_test.go [hunk 10]
```diff
@@ -106,6 +106,7 @@ func TestVerifyObjectStorage(t *testing.T) {
 		{
 			name: "Valid Object Storage Configuration",
 			objectStorage: &ConfigObjectStorage{
+				Location:   "s3",
```
Comment (daveaugustus) [inline lib/config/hadeployconfigverify_test.go:109]:
```suggestion
				Location:   "S3",
```

File: lib/config/hadeployconfigverify_test.go [hunk 11]
```diff
@@ -117,6 +118,7 @@ func TestVerifyObjectStorage(t *testing.T) {
 		{
 			name: "Missing Bucket Name",
 			objectStorage: &ConfigObjectStorage{
+				Location:  "s3",
```
Comment (daveaugustus) [inline lib/config/hadeployconfigverify_test.go:121]:
```suggestion
				Location:  "S3",
```

File: lib/config/hadeployconfigverify_test.go [hunk 12]
```diff
@@ -127,6 +129,7 @@ func TestVerifyObjectStorage(t *testing.T) {
 		{
 			name: "Missing Access Key",
 			objectStorage: &ConfigObjectStorage{
+				Location:   "s3",
```
Comment (daveaugustus) [inline lib/config/hadeployconfigverify_test.go:132]:
```suggestion
				Location:   "S3",
```

File: lib/config/hadeployconfigverify_test.go [hunk 13]
```diff
@@ -137,6 +140,7 @@ func TestVerifyObjectStorage(t *testing.T) {
 		{
 			name: "Missing Secret Key",
 			objectStorage: &ConfigObjectStorage{
+				Location:   "s3",
```
Comment (daveaugustus) [inline lib/config/hadeployconfigverify_test.go:143]:
```suggestion
				Location:   "S3",
```

File: lib/config/hadeployconfigverify_test.go [hunk 14]
```diff
@@ -147,6 +151,7 @@ func TestVerifyObjectStorage(t *testing.T) {
 		{
 			name: "Missing Endpoint",
 			objectStorage: &ConfigObjectStorage{
+				Location:   "s3",
```
Comment (daveaugustus) [inline lib/config/hadeployconfigverify_test.go:154]:
```suggestion
				Location:   "S3",
```

File: lib/config/hadeployconfigverify_test.go [hunk 15]
```diff
@@ -157,6 +162,7 @@ func TestVerifyObjectStorage(t *testing.T) {
 		{
 			name: "Invalid Region",
 			objectStorage: &ConfigObjectStorage{
+				Location:   "s3",
```
Comment (daveaugustus) [inline lib/config/hadeployconfigverify_test.go:165]:
```suggestion
				Location:   "S3",
```

File: lib/config/hadeployconfigverify_test.go [hunk 16]
```diff
@@ -165,6 +171,16 @@ func TestVerifyObjectStorage(t *testing.T) {
 			},
 			expectedError: errors.New("invalid AWS region for S3"),
 		},
+		{
+			name: "Empty Type",
+			objectStorage: &ConfigObjectStorage{
+				Location: "gcs",
```
Comment (daveaugustus) [inline lib/config/hadeployconfigverify_test.go:177]:
```suggestion
				Location: "GCS",
```

**Review by daveaugustus — CHANGES_REQUESTED** [review]

File: lib/config/hadeployconfig.go (hunk 2 above)
Comment (YashviJain01) [reply lib/config/hadeployconfig.go:261]: We should create a constant fot this

File: lib/config/hadeployconfig.go [hunk 17]
```diff
@@ -237,6 +255,23 @@ func (c *HaDeployConfig) Parse(configFile string) error {
...
+			if err != nil {
+				return fmt.Errorf("error reading Json file: %w", err)
+			}
+			err = json.Unmarshal(filepath, &gcpServiceAccount)
+			if err != nil {
+				return fmt.Errorf("error unmarshalling Json file: %w", err)
+			}
+			c.ObjectStorage.Config.GcpServiceAccount = &gcpServiceAccount
+		} else if objectStorageConfig.Location == "" {
+			c.ObjectStorage.Config.Location = "s3"
```
Comment (YashviJain01) [inline lib/config/hadeployconfig.go:272]: for s3 as well

**Comment by SahithiMy** [comment]
Pipeline: https://buildkite.com/chef/chef-automate-main-verify-private/builds/31831

**Review by Dmaddu — CHANGES_REQUESTED** [review]
some of the mandatory checks on the pipelines were not passed. Please fix them.

**Review by daveaugustus — APPROVED** [review]

**Review by Dmaddu — APPROVED** [review]

**Pull request merged**